File: JavaScript/1091_shortestPathBinaryMatrix.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var shortestPathBinaryMatrix = function(grid) {

    // 事前檢查，如果頭尾無法前進直接回傳-
    if ((grid[0][0] != 0) || (grid[grid.length-1][grid.length-1] != 0)) return -1
    // 如果只有一格則直接回傳1
    if (grid.length == 1) return 1

    // 宣告需使用的變數、儲存格
    let num = grid.length
    let nextSteps = [[0, 0, 1]]
    let visited = [[0, 0]]
    let steps = [
        [-1, -1], [-1, 0], [-1, 1],
        [0, -1], [0, 1],
        [1, -1], [1, 0], [1, 1]
    ]

    // 當nextStep中有待訪問座標，則對該座標進行嘗試
    while (nextSteps.length > 0) {
        let row, col, distance;
        [row, col, distance] = nextSteps.shift()
        // console.log(`row = ${row}, col = ${col}, num = ${num}, distance = ${distance}`)
        // 如果是終點，直接回傳結果
        if ((row == num-1) && (col == num-1)) return distance
        // 造訪八個前進方向，並判斷是否已觸碰邊界條件
        for (let step of steps) {
            let nextRow = row + step[0]
            let nextCol = col + step[1]
            if (!((nextRow >= 0) && (nextCol >= 0) && (nextRow < num) && (nextCol < num) && (grid[nextRow][nextCol] == 0))) {
                continue;
            }
            let nextStep = [nextRow, nextCol]
            let isVisited = visited.some(x => nextStep.every((v, i) => v === x[i]))
            if (!(isVisited)) {
                nextSteps.push([nextRow, nextCol, distance+1])
                visited.push([nextRow, nextCol])
            }
        }
        console.log("+"*50)
    }

    return -1;
    
};

/**
 * @param {number[][]} grid
 * @return {number}
 */
var shortestPathBinaryMatrix = function(grid) {

    // 事前檢查，如果頭尾無法前進直接回傳-
    if ((grid[0][0] != 0) || (grid[grid.length-1][grid.length-1] != 0)) return -1
    // 如果只有一格則直接回傳1
    if (grid.length == 1) return 1

    // 宣告需使用的變數、儲存格
    let num = grid.length
    let nextSteps = [[0, 0, 1]]
    let visited = {}
    visited[[0, 0]] = 1
    let steps = [
        [-1, -1], [-1, 0], [-1, 1],
        [0, -1], [0, 1],
        [1, -1], [1, 0], [1, 1]
    ]

    // 當nextStep中有待訪問座標，則對該座標進行嘗試
    while (nextSteps.length > 0) {
        let row, col, distance;
        [row, col, distance] = nextSteps.shift()
        // console.log(`row = ${row}, col = ${col}, num = ${num}, distance = ${distance}`)
        // 如果是終點，直接回傳結果
        if ((row == num-1) && (col == num-1)) return distance
        // 造訪八個前進方向，並判斷是否已觸碰邊界條件
        for (let step of steps) {
            let nextRow = row + step[0]
            let nextCol = col + step[1]
            if (!((nextRow >= 0) && (nextCol >= 0) && (nextRow < num) && (nextCol < num) && (grid[nextRow][nextCol] == 0))) {
                continue;
            }
            if (!(visited.hasOwnProperty([nextRow, nextCol]))) {
                nextSteps.push([nextRow, nextCol, distance+1])
                visited[[nextRow, nextCol]] = 1
            }
        }
        console.log("+"*50)
    }

    return -1;
    
};

/**
 * @param {number[][]} grid
 * @return {number}
 */
var shortestPathBinaryMatrix = function(grid) {

    // 事前檢查，如果頭尾無法前進直接回傳-
    if ((grid[0][0] != 0) || (grid[grid.length-1][grid.length-1] != 0)) return -1
    // 如果只有一格則直接回傳1
    if (grid.length == 1) return 1

    // 宣告需使用的變數、儲存格
    let num = grid.length
    let nextSteps = [[0, 0]]
    let distance = 1
    let steps = [
        [-1, -1], [-1, 0], [-1, 1],
        [0, -1], [0, 1],
        [1, -1], [1, 0], [1, 1]
    ]

    // 當nextStep中有待訪問座標，則對該座標進行嘗試
    while (nextSteps.length > 0) {
        let numOfSteps = nextSteps.length;
        for (let i=0; i<numOfSteps; i++) {
            let row, col;
            [row, col] = nextSteps.shift()
            // 如果是終點，直接回傳結果
            if ((row == num-1) && (col == num-1)) return distance
            // 造訪八個前進方向，並判斷是否已觸碰邊界條件
            for (let step of steps) {
                let nextRow = row + step[0]
                let nextCol = col + step[1]
                if (((nextRow >= 0) && (nextCol >= 0) && (nextRow < num) && (nextCol < num) && (grid[nextRow][nextCol] == 0))) {
                    grid[nextRow][nextCol] = 1
                    nextSteps.push([nextRow, nextCol, distance+1])
                }
            }
        }
        distance += 1
    }

    return -1;
};

console.log(shortestPathBinaryMatrix([[0,1],[1,0]]))
console.log(shortestPathBinaryMatrix([[0,0,0],[1,1,0],[1,1,0]]))
console.log(shortestPathBinaryMatrix([[1,0,0],[1,1,0],[1,1,0]]))

File: Python/1091_shortestPathBinaryMatrix.py
from typing import List


class Solution2:
    '''BFS
        Time: O(n^2)
        Space: O(n^2)
    '''
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        
        if grid[0][0] == 1 or grid[-1][-1] == 1: return -1

        num = len(grid) # length of the grid
        nextSteps = [[0, 0]] # array used to record next direction
        distance = 1
        steps = [
            [-1, -1], [-1, 0], [-1, 1],
            [0, -1], [0, 1],
            [1, -1], [1, 0], [1, 1]
        ] # eight directions can move

        grid[0][0] = 1
        while nextSteps:
            for _ in range(len(nextSteps)):
                row, col = nextSteps.pop(0)
                # return conditin
                # 因為 BFS 特性，可以在這裡提早返回，因為已找到最短距離
                if row == num-1 and col == num-1: return distance
                for step in steps:
                    next_row = row + step[0]
                    next_col = col + step[1]
                    # Skip this step if out of the grid
                    if next_row >= 0 and next_col >= 0 and next_row < num and next_col < num and grid[next_row][next_col] == 0:
                        grid[next_row][next_col] = 1
                        nextSteps.append([next_row, next_col])
            distance += 1
        
        return -1


class Solution3:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        """
        DFS
            Time: O(2^{n^2})
            Space: O(n^2)
        """

        # 如果起點為(1,1)，則代表無可行道路
        if grid[0][0] == 1: return -1

        M, N = len(grid), len(grid[0])
        directions = [
            [-1, -1], [-1, 0], [-1, 1],
            [0, -1], [0, 1],
            [1, -1], [1, 0], [1, 1]
        ]
        min_cnt = float('inf')

        def dfs(m, n, cnt):
            if m == M-1 and n == N-1:
                nonlocal min_cnt
                min_cnt = min(cnt, min_cnt)
                return
            for d in directions:
                x, y = d[0], d[1]
                if 0 <= m+x < M and 0 <= n+y < N and grid[m+x][n+y] == 0:
                    grid[m][n] = 1 # 將走過的路標為1
                    dfs(m+x, n+y, cnt+1)
                    grid[m][n] = 0 # 走完後再標回0
        grid[0][0] = 1
        dfs(0, 0, 1)
        return min_cnt if min_cnt != float('inf') else -1

        

if __name__ == '__main__':
    # print(Solution3().shortestPathBinaryMatrix([[0,1],[1,0]])) # 2
    print(Solution3().shortestPathBinaryMatrix([[0,0,0],[1,1,0],[1,1,0]])) # 4
    # print(Solution3().shortestPathBinaryMatrix([[1,0,0],[1,1,0],[1,1,0]])) # -1